aten/src/ATen/core/ivalue.h

```diff
@@ -186,6 +187,10 @@ struct CAFFE2_API IValue final {
     return payload.as_double;
   }
 
+  IValue(ScalarType t) : tag(Tag::ScalarType) {
+    payload.as_int = static_cast<std::underlying_type<ScalarType>::type>(t);
+  }
+
```

Comment:
naively, I would also expect an isScalarType and that toScalarType actually checks the tag is a scalar type.

I'm admittedly not sure if we need to deal with BC constraints here, though.  @suo do you know?

Reply:
this was wrong, will fix

Reply:
added the is_intrusive_ptr, isScalarType, and added the check in toScalarType

Reply:
ok, this looks correct now, but I don't know about the BC concerns.

Reply:
what is it that would potentially backwards incompatible?

Reply:
my concern was you serialized something as a tagged-int previously and when you load it in the new version it errors out.

Reply:
hmm I do have a test failing here now.
`python test_cpp_api_parity.py TestCppApiParity.test_torch_nn_SampleModule_has_parity` gives "ivalue not a ScalarType" and I assume that's because it's trying to deserialize an int as a scalar type.

Reply:
removed the isScalarType check. toLayout, toMemoryFormat, etc don't have it and it was causing test failures.

Reply:
For correctness, adding a new tag requires adding a corresponding type in jit_type.h, and adjusting the schema/type parsers to understand that 'ScalarType' maps to this type. This PR seems to just add the tag and not add a type. If we need to keep the tag around, then we should also add the corresponding type and fix the type parser.

Reply:
@zdevito this PR does add a new type in jit_type.h, is it incorrect?

Reply:
https://github.com/pytorch/pytorch/pull/26012/files#diff-7e8e36dc4eff19e3d5d1daa977da20eeR1142

Reply:
removed almost all jit changes